File: lecture063/minStack.java
class MinStack {

    Stack<Integer> st;
    Stack<Integer> minStack;
    int mini = Integer.MAX_VALUE;

    /** initialize your data structure here. */
    public MinStack() {
        st = new Stack<>();
        minStack = new Stack<>();
    }

    public void push(int x) {
        st.push(x);
        minStack.push(Math.min(mini, x));
        mini = minStack.peek();
    }

    public void pop() {
        if (st.size() == 0)
            return;

        st.pop(x);
        minStack.pop(x);

        if (minStack.size() != 0) {
            min = minStack.peek();
        }
        else {
            min = Integer.MAX_VALUE;
        }
    }

    public int top() {
        return st.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}

class MinStack2 {
    Stack<Long> st;
    long minsf = 0;
    public MinStack2() {
        st = new Stack<>();
    }

    public void push(int x) {
        long val = x;
        if (st.size() == 0) {
            minsf = val;
            st.push(val);
            return;
        }

        if (val > minsf) {
            st.push(val);
        }
        else {
            st.push(val - minsf + val);
            minsf = val;
        }
    }

    public int top() {
        if (st.peek() <= minsf) {
            return (int)minsf;
        }
        long val = st.peek();
        return (int) val; 
    }

    public void pop() {
        if (st.peek() < minsf) {
            minsf = minsf - st.peek() + minsf;
        }
        st.pop();
    }

    public int getMin() {
        return (int)minsf;
    }
}

File: lecture062/minStack.cpp

// using 2 stacks

class MinStack {
public:
    /** initialize your data structure here. */
    int minsf = INT_MAX;
    stack<int> st;
    stack<int> minSt;
    MinStack() {
        
    }
    
    void push(int x) {
        st.push(x);
        minSt.push(min(minsf, x));
        minsf = minSt.top();
    }
    
    void pop() {
        if (st.size() == 0)
            return;
        
        st.pop();
        minSt.pop();
        
        if(minSt.size() != 0)
            minsf = minSt.top();
        else 
            minsf = INT_MAX;
    }
    
    int top() {
        return st.top();
    }
    
    int getMin() {
        return minSt.top();
    }
};

//using 1 stack (better)

class MinStack {
public:
    /** initialize your data structure here. */
    long minsf = INT_MAX;
    stack<long> st;
    MinStack() {
        
    }
    
    void push(int x) {
        long value = x;
        if (st.size() == 0) {
            st.push(value);
            minsf = value;
            return;
        }
        
        if (value > minsf)
        {
            st.push(value);
        }
        else 
        {
            st.push(value - minsf + value);
            minsf = value;
        }
    }
    
    void pop() {
        if (st.top() < minsf)
            minsf = minsf - st.top() + minsf;
        st.pop();
    }
    
    int top() {
        if (st.top() <= minsf)
            return (int)minsf;
        
        return (int)st.top();
    }
    
    int getMin() {
        return (int)minsf;
    }
};